test/e2e/storage/testsuites/volumelimits.go

```diff
@@ -0,0 +1,335 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	migrationplugins "k8s.io/csi-translation-lib/plugins" // volume plugin names are exported nicely there
+	volumeutil "k8s.io/kubernetes/pkg/volume/util"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+)
+
+type volumeLimitsTestSuite struct {
+	tsInfo TestSuiteInfo
+}
+
+const (
+	// The test uses generic pod startup / PV deletion timeouts. As it creates
+	// much more volumes at once, these timeouts are multiplied by this number.
+	// Using real nr. of volumes (e.g. 128 on GCE) would be really too much.
+	testSlowMultiplier = 10
+
+	// How long to wait until CSINode gets attach limit from installed CSI driver.
+	csiNodeInfoTimeout = 1 * time.Minute
+)
+
+var _ TestSuite = &volumeLimitsTestSuite{}
+
+// InitVolumeLimitsTestSuite returns volumeLimitsTestSuite that implements TestSuite interface
+func InitVolumeLimitsTestSuite() TestSuite {
+	return &volumeLimitsTestSuite{
+		tsInfo: TestSuiteInfo{
+			name: "volumeLimits",
+			testPatterns: []testpatterns.TestPattern{
+				testpatterns.FsVolModeDynamicPV,
+			},
+		},
+	}
+}
+
+func (t *volumeLimitsTestSuite) getTestSuiteInfo() TestSuiteInfo {
+	return t.tsInfo
+}
+
+func (t *volumeLimitsTestSuite) defineTests(driver TestDriver, pattern testpatterns.TestPattern) {
+	type local struct {
+		config      *PerTestConfig
+		testCleanup func()
+
+		cs clientset.Interface
+		ns *v1.Namespace
+		// genericVolumeTestResource contains pv, pvc, sc, etc. of the first pod created
+		resource *genericVolumeTestResource
+
+		// All created PVCs, incl. the one in resource
+		pvcs []*v1.PersistentVolumeClaim
+
+		// All created PVs, incl. the one in resource
+		pvNames sets.String
+
+		runningPod       *v1.Pod
+		unschedulablePod *v1.Pod
+	}
+	var (
+		l local
+	)
+
+	// No preconditions to test. Normally they would be in a BeforeEach here.
+	f := framework.NewDefaultFramework("volumelimits")
+
+	// This checks that CSIMaxVolumeLimitChecker works as expected.
+	// A randomly chosen node should be able to handle as many CSI volumes as
+	// it claims to handle in CSINode.Spec.Drivers[x].Allocatable.
+	// The test uses one single pod with a lot of volumes to work around any
+	// max pod limit on a node.
+	// And one extra pod with a CSI volume should get Pending with a condition
+	// that says it's unschedulable because of volume limit.
+	// BEWARE: the test may create lot of volumes and it's really slow.
+	ginkgo.It("should support volume limits [Slow][Serial]", func() {
+		driverInfo := driver.GetDriverInfo()
+		if !driverInfo.Capabilities[CapVolumeLimits] {
+			ginkgo.Skip(fmt.Sprintf("driver %s does not support volume limits", driverInfo.Name))
+		}
+		var dDriver DynamicPVTestDriver
+		if dDriver = driver.(DynamicPVTestDriver); dDriver == nil {
```

Comment:
I thought these checks were all done in some helper function somewhere thats shared between test suites

Reply:
Technically the error path is not reachable. Retyping `driver` to `DynamicPVTestDriver` should always succeed, but I have a safety check just to be sure.